### ISSUE TITLE
SHACL: add String-based constraints

### ISSUE THREAD

**Issue by barthanssens** [eclipse/rdf4j-storage#49]
SHACL 4.4, most of the string-based constraints are probably fairly easy to add: minLength / maxLength / pattern / languageIn

(uniqueLang is more difficult / resource intensive)

See also https://www.w3.org/TR/shacl/#core-components-string

**Comment by hmottestad** [comment]
Datatype based restrictions are also useful and as simple as string based ones.

**Comment by hmottestad** [comment]
The plan for validating such data should kick in if:
 - a node of type targetClass is added
 - a statement that matches the path is added

The plan should handle the following cases:
1. data added, type statement and path in added
2. data added, path in added, type statement in underlying store
3. data added, type statement added, path in underlying store

Note that it is possible with combinations of these cases. 

Case 1 can easily degrade into case 3 by adding in the statements matching path for added nodes of type targetClass. This can then degrade into case 2 by filtering on the type statement from the underlying store.

**Comment by barthanssens** [comment]
Indeed. I think I've managed the string constraints working, but probably suboptimal (seem to run more than they should)

**Comment by hmottestad** [comment]
That's great. Do you have a branch?

**Comment by barthanssens** [comment]
https://github.com/fedict/rdf4j-storage/tree/issues/%2349-shacl-stringconstraints

Still pretty much work in progress, I'll check this evening if the code on github is actually the latest version of the I'm working on.
I did some renaming and code rework (factories in separate classes e.g).

Currently I'm checking minLength,maxLength and pattern in 1 single check, although I'm not sure that is the best option - or at least it may require additional return values -  if there is a need to also implement the sh:ValidationReport reports.

**Comment by hmottestad** [comment]
I had a quick look now. Lots of nice cleanup :)

A couple of things I would like to keep though. Factory classes in the classes they belong to, makes it easier to understand and find, having them outside doubles the number of files and makes it harder to figure out which factory is for which class. Also a pain when renaming things later. Also, the queries with the binds, point here was to be able to track which statements in the database were used in evaluating a shacl rule, so that we can give an “explanation”. Currently this is being logged.

Try to make some tests for your added rules. Remember that shacl rules with target class trigger for two cases, when an rdf:type statement is added or when a statement matching the path is added (for min count also when removed, but that’s not applicable here).

**Comment by barthanssens** [comment]
Thanks for the review :-)
I'll move the factories back into the original file, and I'll try to keep working on it (if time permits...)
